litestar/channels/redis.py

```diff
@@ -0,0 +1,126 @@
+        super().__init__(history, redis=redis)
+        self._pub_sub: PubSub = self._redis.pubsub()
+
+    async def on_startup(self) -> None:
+        pass
+
+    async def on_shutdown(self) -> None:
+        await self._pub_sub.reset()
+
+    async def subscribe(self, channels: Iterable[str]) -> None:
+        await self._pub_sub.subscribe(*channels)
+
+    async def unsubscribe(self, channels: Iterable[str]) -> None:
+        await self._pub_sub.unsubscribe(*channels)
+
+    async def publish(self, data: bytes, channels: Iterable[str]) -> None:
+        for channel in channels:
+            await self._redis.publish(channel, data)
+
+    async def _wait_for_subscribed_channels(self) -> None:
+        while True:
+            if self._pub_sub.channels:
+                break
+            await asyncio.sleep(0)
+
+    async def stream_events(self) -> AsyncGenerator[tuple[str, Any], None]:
+        if not self._pub_sub:
+            raise RuntimeError()
+
+        while True:
+            await self._wait_for_subscribed_channels()
+            message = await self._pub_sub.get_message(ignore_subscribe_messages=True)
+            if message is None:
+                continue
+            channel = message["channel"].decode()
+            data = message["data"]
+            yield channel, data
+
+    async def get_history(self, channel: str, limit: int | None = None) -> list[bytes]:
+        raise NotImplementedError()
+
+
+class RedisChannelsStreamBackend(RedisChannelsBackend):
+    def __init__(self, history: int = 0, *, redis: Redis, cap_streams_approximate: bool = True) -> None:
+        super().__init__(history, redis=redis)
+        if history < 1:
+            raise ImproperlyConfiguredException("history must be greater than 0")
+
+        self._subscribed_channels: set[str] = set()
+        self._cap_streams_approximate = cap_streams_approximate
+
+    async def on_startup(self) -> None:
+        pass
+
+    async def on_shutdown(self) -> None:
+        pass
+
+    async def subscribe(self, channels: Iterable[str]) -> None:
+        self._subscribed_channels.update(channels)
+
+    async def unsubscribe(self, channels: Iterable[str]) -> None:
+        self._subscribed_channels = self._subscribed_channels - set(channels)
+
+    async def publish(self, data: bytes, channels: Iterable[str]) -> None:
+        for channel in channels:
+            await self._redis.xadd(
+                self._make_key(channel),
+                {"data": data, "channel": channel},
+                maxlen=self._history,
+                approximate=self._cap_streams_approximate,
+            )
+
+    async def stream_events(self) -> AsyncGenerator[tuple[str, Any], None]:
+        stream_ids: dict[str, bytes] = {}
+        while True:
+            stream_keys = [self._make_key(c) for c in self._subscribed_channels]
+            if not stream_keys:
+                await asyncio.sleep(0)
+                continue
```

Comment:
This is not ideal, but the other solutions (e.g. using a `Future` or `Event`) would require a lock around `_subscribed_channels` to ensure their contents stay the same once execution resumes.

An alternative would be to use both an `Event` and still do the check as a fallback. This should lead to the event blocking in most cases where no channels are subscribed and the condition with `sleep(0)` should only be hit in case of a race conditions. @litestar-org/maintainers thoughts?